(maint) add grouped jobs concept and count interface

jonathannewman opened the pull request:
This adds the concept of jobs within the same group.  Jobs that are grouped
can be counted or stopped together.  This is helpful when managing jobs associated
with a given trapperkeeper service.

Upon service stop, the jobs set is now emptied.  Previously the jobs were stopped
but the set maintained the entries in the set.

`after` calls are now given a unique identifier and removed from the set when they
complete.

jonathannewman commented:
@KevinCorcoran (and whomever else is interested), I was trying to use the scheduler service and found some areas that would be helpful for the apps I've been integrating the service into.  I added a few interfaces, and added some additional tracking for `after` jobs, and cleaned up the job set when the service is shutdown.

KevinCorcoran commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_service.clj:
```diff
@@ -19,6 +19,33 @@
+  (if-let [item (first (filter #(= id (:id %)) @(get-jobs this)))]
+    (swap! (get-jobs this) disj item)))
+
+(defn- wrap-remove-job-by-after-id!
```
It took me a while to understand what was up with this, but I think I finally get it (mostly).  But I have a handful of comments:

* Most importantly, why do we need a separate `after-id`? It seems like we could reuse the existing ID for this ... but maybe I'm not understanding something about how `after-id` is used.

*  Does this mean that, before these changes, `after` job IDs were only added to the list of jobs, but never removed?

* Finally, I struggled with this name.  I think that, if this function had a docstring, a simple name like `after-job` might be sufficient.

jonathannewman replied:
My intent was to prevent memory leaks with the "after" jobs identifiers.  They will remain in the job set after the job completes, if they are not explicitly removed (which is what I'm doing in the finally block).   There is a chicken-and-egg issue with using the existing job identifier -- it doesn't get created until after the job is scheduled.    I'll add some docs strings and play with the naming, maybe that will make it clearer.

KevinCorcoran commented on test/integration/puppetlabs/trapperkeeper/services/scheduler/scheduler_service_test.clj:
```diff
@@ -7,53 +7,64 @@
-                  (Thread/sleep 100)
+  (let [configurations [{}
+                        {:group-id :some-group-identifier}]]
+    (doseq [{:keys [group-id]} configurations]
```
I prefer to not use `doseq` in situations like this because it makes it difficult to debug test failures as it's usually not clear which iteration(s) a given assertion failed on.  Duplicating the needed parts of the test makes the code WETter but much easier to debug when tests fail.

jonathannewman replied:
Ok.

KevinCorcoran commented on test/integration/puppetlabs/trapperkeeper/services/scheduler/scheduler_service_test.clj:
```diff
@@ -62,37 +73,185 @@
   (set
     (for [x [1 2 3]]
-      (interspaced service 1000 (constantly x)))))
+      (:id (interspaced service 1000 (constantly x))))))
```
What's going on here?  Did this change the return value of `interspaced`?

jonathannewman replied:
The job identifier is now different but that should be transparent to users... this test change was unintentional, I'll remove it.

jonathannewman replied:
Yes, the interface functions no longer just return the `at/at` job directly.  They are now wrapped in a map that contains optionally the `group-id` and maybe `after-id`.  The job is under the `id` key now, so this is returning the `at/at` job, which the test below expects.

KevinCorcoran replied:
Okay, that makes sense.  It seems a little strange that the at-at job is referred to as `:id` but I realize that's consistent with how things were done previously, so 👍 

This is an API change, so if this project was post-1.0, it would be a major version bump.  Given that it's still pre-1.0, perhaps it would just be best to bump the project version to `0.2.0-SNAPSHOT` as a part of this PR, so that the next release is at least a Y.

haus replied:
The currently released version is 0.0.1 and the next version as currently written in project.clj will be 0.1.0. I'm not sure moving to 0.2.0 instead of 0.1.0 matters much here.

KevinCorcoran commented on test/integration/puppetlabs/trapperkeeper/services/scheduler/scheduler_service_test.clj:
```diff
@@ -62,37 +73,185 @@
+            (Thread/sleep 100)
+            (is (= 1 (count-jobs service)))
+            (stop-job service job-2)
+            (Thread/sleep 100)
```
`stop-job` is synchronous, so these `sleeps` are unnecessary.  Or am I misunderstanding something?

Oh, it's because the jobs might be running when `stop-job` is called, isn't it?  It'd be worth adding a comment about that.

jonathannewman replied:
Yep.  Will do.

KevinCorcoran commented on test/integration/puppetlabs/trapperkeeper/services/scheduler/scheduler_service_test.clj:
```diff
@@ -62,37 +73,185 @@
+        (is (= 1 (count-jobs service)))
+        (deliver wait-for-start true)
+        (deref completed)
+        (Thread/sleep 100)
```
Okay, this `sleep` REALLY seems unnecessary.  :)

(since this is using `after` instead of `interspaced`)

jonathannewman replied:
The job could have completed, but not removed from the set. There is a small window (that finally block in the "after" function) that needs to get executed before it is really removed, so I think it is needed.

KevinCorcoran commented on test/integration/puppetlabs/trapperkeeper/services/scheduler/scheduler_service_test.clj:
```diff
@@ -62,37 +73,185 @@
     (with-app-with-empty-config app [scheduler-service]
       (let [service (tk/get-service app :SchedulerService)
             started (promise)
             stopped (promise)
```
`stopped` is unused now

KevinCorcoran commented on test/integration/puppetlabs/trapperkeeper/services/scheduler/scheduler_service_test.clj:
```diff
@@ -62,37 +73,185 @@
+            group-id-0 :unique-group-id
+            group-id-1 :more-unique-group-id
+            ; create one job without a service id and two with one group-id
+            ; and a third with a different service id
```
I think you mean `group id`, not `service id`.  This applies to a few places below as well.

jonathannewman replied:
Yes, I changed my mind when writing this.  I thought I caught them all, I'll update.

KevinCorcoran commented on test/integration/puppetlabs/trapperkeeper/services/scheduler/scheduler_service_test.clj:
```diff
@@ -7,53 +7,64 @@
+                      (if group-id #(after service delay job group-id)
+                                   #(after service delay job))]
+                (let [schedule-time (System/currentTimeMillis)]
+                  (after service delay job)
```
looks like this should be `after-fn`

jonathannewman commented on src/puppetlabs/trapperkeeper/services/protocols/scheduler.clj:
```diff
@@ -2,17 +2,37 @@
-    Returns 'true' if the job was successfully stopped, 'false' otherwise."))
+    Returns 'true' if the job was successfully stopped, 'false' otherwise.")
+
+  (stop-grouped-jobs [this group-id]
```
Sounds good.

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_service.clj:
```diff
@@ -30,27 +57,49 @@
     (log/debug "Shutting down Scheduler Service")
     ; Stop any jobs that are still running
     (core/stop-all-jobs! @(:jobs context) (get-pool this))
+    (reset! (:jobs context) #{})
```
Now that we have "reset" behavior with tk, it looked like the job identifiers wouldn't get cleaned up properly on stop (if they were left running), which could result in a memory leak.

KevinCorcoran replied:
Since the `init` function does this ...
```clj
(assoc context :pool pool
                     :jobs (atom #{})
                     :after-id (atom 0))
```
I don't think this line is strictly necessary or a memory leak.  But no big deal, I don't mind leaving it, just wanted to understand.

jonathannewman replied:
Oh, I missed that! I'll remove the addition.

jonathannewman commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_service.clj:
```diff
@@ -30,27 +57,49 @@
+  (stop-grouped-jobs [this group-id]
+    (let [jobs-by-group (jobs-by-group-id this group-id)]
+       (doall
+         (for [job jobs-by-group]
```
Interesting.  I'll change it.

jonathannewman commented:
@KevinCorcoran I made the suggested changes and was able to remove almost all of the sleeps.

KevinCorcoran commented on test/integration/puppetlabs/trapperkeeper/services/scheduler/scheduler_service_test.clj:
```diff
@@ -62,37 +114,195 @@
+          (is (= 0 (count-jobs service group-id-0)))
+          (is (= 1 (count-jobs service group-id-1))))
+
+        (testing "stopping one job does not stop the service id based job"
```
still a couple `service id` usages here

jonathannewman commented:
@KevinCorcoran will update snapshot version and fix the `service id` usages.

jonathannewman commented:
I changed the `:id` reference into `:job`, removed the two straggling service id references.  In addition, I added another arity to `dequeue-job!` as we we doing redundant work in the case we were actually given the job to remove.

haus commented on src/puppetlabs/trapperkeeper/services/scheduler/scheduler_service.clj:
```diff
@@ -19,6 +19,45 @@
+    (swap! (get-jobs this) disj job))
+  ([this id keyword]
+    (when-let [item (first (filter #(= id (get % keyword)) @(get-jobs this)))]
+    (swap! (get-jobs this) disj item))))
```
nit: indentation looks a little off on the `swap!` line

haus commented:
I had a single nit-pick. I'm 👍  here.

jonathannewman commented:
Fixed the alignment problem.
